src/main/java/core/basesyntax/ArrayList.java

```diff
@@ -1,48 +1,141 @@
+                }
+            }
+        } else {
+            grow();
+            internalArray[size] = value;
+            size++;
+        }
     }
 
     @Override
     public void add(T value, int index) {
-
+        if (index >= 0 && index <= size) {
+            growByCell(index);
+            internalArray[index] = value;
+            size++;
+        } else {
+            throw new ArrayListIndexOutOfBoundsException("Wrong index!!!");
+        }
     }
 
     @Override
     public void addAll(List<T> list) {
-
+        for (int i = 0; i < list.size(); i++) {
+            add(list.get(i));
+        }
     }
 
     @Override
     public T get(int index) {
-        return null;
+        if (index < size && index >= 0) {
+            return (T) this.internalArray[index];
+        } else {
+            throw new ArrayListIndexOutOfBoundsException("The such index doesn't exist");
+        }
     }
 
     @Override
     public void set(T value, int index) {
-
+        if (index < size && index >= 0) {
+            internalArray[index] = value;
+        } else {
+            throw new ArrayListIndexOutOfBoundsException("Wrong array index!");
+        }
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        T oldelement = null;
+        if (index < this.size && index >= 0) {
+            int subsiquentIndex = 0;
+            oldelement = get(index);
+            Object[] newFormedArray = new Object[internalArray.length];
+            for (int i = 0; i < size; i++) {
+                if (i != index) {
+                    newFormedArray[subsiquentIndex] = internalArray[i];
+                    subsiquentIndex++;
+                }
+            }
+            internalArray = newFormedArray;
+            size--;
+        } else {
+            throw new ArrayListIndexOutOfBoundsException("Wrong index");
+        }
+        return oldelement;
     }
 
     @Override
     public T remove(T element) {
-        return null;
+        T oldElement = null;
+        for (int i = 0; i < size; i++) {
+            if (internalArray[i] != null && internalArray[i].equals(element)) {
+                oldElement = element;
+                remove(i);
+                break;
+            } else if (internalArray[i] == null && internalArray[i] == element) {
+                remove(i);
+                break;
+            }
+        }
+        if (oldElement == null && element != null) {
+            throw new NoSuchElementException("This element hasn't found");
+        }
+        return oldElement;
     }
 
     @Override
     public int size() {
-        return 0;
+        return this.size;
     }
 
     @Override
     public boolean isEmpty() {
-        return false;
+        return size == 0;
+    }
+
+    private void grow() {
+        Object[] bufferArray = new Object[size + (size >> 1)];
+        System.arraycopy(internalArray, 0, bufferArray, 0, size);
+        internalArray = bufferArray;
+    }
```

Comment:
The 'grow' method should be responsible only for increasing the size of the internal array. Move the logic of copying the array elements into a separate method to adhere to the Single Responsibility Principle.